### PR TITLE
Skip native builds - part 2

### PR DESCRIPTION
**Pull request by hawkrives**
followup to #3209 

part of #3250? maybe?

now with actual build skipping!

I pulled the should-build script out into a separate Ruby file - it could probably stand some poking, @rye.

- [x] ~I should go make it print '0' to build-status, because Danger's still going to want that~ done

image

### CODE REVIEW

**Comment by hawkrives** [comment]
@rye meh - went another route :p

**Comment by hawkrives** [comment]
(thanks though)

**Comment by hawkrives** [comment]
So! Reviews? @rye, @hannesmcman, @drewvolz?

This seems to work!

File: scripts/should-build.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    puts('should_build? some native file changed, so yes')
+    return true
+  end
+
+  # 4) if none matched, return false
+  puts('should_build? no')
+  false
+end
+
+def main
```
Comment (rye) [inline scripts/should-build.rb]: This method is kinda pointless.  Consider just putting the body outside of the function (at the top level?)

File: scripts/should-build.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
+require 'json'
+
+PLATFORM = ENV.fetch('task', 'unknown')
```
Comment (rye) [inline scripts/should-build.rb]: I prefer the following syntax:
```suggestion
PLATFORM = ENV['task'] || 'unknown'
```

Also, freeze it?

```suggestion
PLATFORM = (ENV['task'] || 'unknown').freeze
```

File: scripts/should-build.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+def git_branch
+  sh('git rev-parse --abbrev-ref HEAD')
+end
+
+# diff two hashes at a path and return the changed items
+def deps_diff(old_hash, new_hash, path)
+  (new_hash.dig(*path).to_a - old_hash.dig(*path).to_a).to_h
+end
+
+NPM_DEP_NAME_REGEXP = /react|jsc/.freeze
```
Comment (rye) [inline scripts/should-build.rb:21]: Can we get some whitespace between this and the method that follows?

File: scripts/should-build.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  end
+
+  # 4) if none matched, return false
+  puts('should_build? no')
+  false
+end
+
+def main
+  if should_build?
+    File.open(File.join(File.dirname(__FILE__), '..', 'logs', 'build-status'), 'w') { |file| file.write('0') }
```
Comment (rye) [inline scripts/should-build.rb]: I would imagine RuboCop wants this line truncated.

Maybe use `do...end` syntax instead?

**Review by rye — CHANGES_REQUESTED** [review]
Just a couple of stylistic nits

File: scripts/should-build.rb (hunk 2 above)
Comment (hawkrives) [reply scripts/should-build.rb]: Rubocop said something about `Do not freeze immutable objects, as freezing them has no effect.` on the regexp one… I can still freeze this one too

Comment (rye) [reply scripts/should-build.rb]: Oh, no need to do that.  We should not freeze the regexp literal then.

Comment (hawkrives) [reply scripts/should-build.rb]: Oh, that went away after upgrading rubocop.

Comment (rye) [reply scripts/should-build.rb]: Oh, then we should freeze it this one too :P

File: scripts/should-build.rb (hunk 3 above)
Comment (hawkrives) [reply scripts/should-build.rb:21]: are you sure we can't just inline this? I really want to inline this again… so concise…

Comment (rye) [reply scripts/should-build.rb:21]: Yes, I'm sure.  We print it out elsewhere.

Comment (hawkrives) [reply scripts/should-build.rb:21]: although then we can't inline it into the other string… bah.

**Review by rye — APPROVED** [review]
Looks good to me!

**Comment by hawkrives** [comment]
(cancelled the last two commits to speed things along)

**Comment by hawkrives** [comment]
https://status.circleci.com/incidents/vvdpb19g0n21 – something broke inside Circle, so the macOS build is stuck for a while.

**Pull request merged**